Feeder subsystem

AbhiJ2706 opened the pull request:
Adding the code for the feeder subsystem. This contains the feeder subsystem code, commands to run the feeder and increment the feeder, and mapped these commands to joystick buttons. 

Further changes will need to be made to configure it for the 2020 robot, and once this code has been tested.

Issue #34

AbhiJ2706 committed:
Added the indexer subsystem

Added the indexer subsystem

AbhiJ2706 committed:
Added commands and documented

Added commands for the feeder subsystem, added necessary lines to the config and Robot.java. Documented all the added code.

georgetzavelas commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -64,6 +67,12 @@ private void configureButtonBindings() {
+    incrementFeederCommand = new IncrementFeeder();
+    new JoystickButton(controlStick, XboxController.Button.kA.value).whenPressed(incrementFeederCommand);
+
+    emptyFeederCommand = new IncrementFeeder();
```
Do you want this to reference `EmptyFeeder`?

kevlam2706 commented on src/main/java/frc/robot/config/FluidConstant.java:
```diff
@@ -76,14 +76,6 @@ public T getValue() {
-    public T get() {
-        return this.value;
-    }
-    
```
Was this block deleted intentionally?

georgetzavelas replied:
This block seems to be deleted in many of the open PRs so I suspect it may be occurring automatically by the IDE as it is possibly unused?

AbhiJ2706 replied:
When I opened the file there were 2 of that method and VSCode was giving an error.

kevlam2706 commented on src/main/java/frc/robot/subsystems/FeederSubsystem.java:
```diff
@@ -0,0 +1,130 @@
+
+    //How much to shift the feeder wheel when incrementing
+    public static FluidConstant<Double> INCREMENT_TICKS = new FluidConstant<>("IncrementTicks", 0.0)
+                .registerToTable(Config.constantsTable);
```
maybe name this with a FeederSubsystem prefix so it is clear what it is?

kevlam2706 commented on src/main/java/frc/robot/subsystems/FeederSubsystem.java:
```diff
@@ -0,0 +1,130 @@
+    public static FluidConstant<Double> INCREMENT_TICKS = new FluidConstant<>("IncrementTicks", 0.0)
+                .registerToTable(Config.constantsTable);
+    //Max distance at which the robot knows a ball is at the indexer
+    public static FluidConstant<Integer> IR_MAX_DISTANCE = new FluidConstant<>("IrMaxDistance", 0)
```
maybe name with a FeederSubsystem prefix so it's clear what it is?

kevlam2706 commented on src/main/java/frc/robot/subsystems/FeederSubsystem.java:
```diff
@@ -0,0 +1,130 @@
+        int absolutePosition = feederTalon.getSensorCollection().getPulseWidthPosition();
+        feederTalon.setSelectedSensorPosition(absolutePosition, kPIDLoopIdx, kTimeoutMs);
+        feederTalon.setSelectedSensorPosition(0, 0, 10);
+
```
Some of these calls return errorCodes - most critically the call to configSelectedFeedbackSensor(). We need to catch the error code and do something about anything abnormal, such as if the encoder isn't found by the Talon.

Nathan has written a cool conditions framework that lets you set a condition called "encoderFound" or "encoderHealthy" or whatever you want to call it. The condition framework allows the subsystem to determine if it should run or not based on every condition being met. You could use that to essentially flag the subsystem as being unable to run if the errorCode is not happy. We should also log the errors.

kevlam2706 reviewed (approved):
Looks good. Look into incorporating the Subsystem Conditions framework that Nathan set up as I think it would be a good way to catch error states and prevent the system from doing anything if, say, the encoder is disconnected or not found.

AbhiJ2706 committed:
Fixed some issues

Fixed some of the issues noted in the pull request. Still have some to fix but I need to pull from master.

AbhiJ2706 committed:
Merge branch 'master' into indexer-subsystem

AbhiJ2706 committed:
Merge branch 'master' into indexer-subsystem

AbhiJ2706 committed:
Fixed rest of issues

Added conditions to the feeder subsystem.

AbhiJ2706 committed:
Fixed type

AbhiJ2706 committed:
Fixed type

LaraLim reviewed (approved):
Looks good.

AbhiJ2706 committed:
Merge branch 'master' into indexer-subsystem
